src/cache.ts

```diff
@@ -0,0 +1,221 @@
+			if (value.startsWith('u8-') || value.startsWith('date-')) {
+				throw new Error('serialization error');
+			}
+		}
+		return value;
+	});
+};
+
+const deserialize = <T>(value: string): T => {
+	return JSON.parse(value, (_key, value) => {
+		if (value.startsWith && value.startsWith('u8-')) {
+			return b64Tou8(value.slice('u8-'.length));
+		}
+		if (value.startsWith && value.startsWith('date-')) {
+			return new Date(value.slice('date-'.length));
+		}
+		return value;
+	});
+};
+
+export class InMemoryCache implements ReadableCache {
+	private store: Map<string, CacheElement<string>>;
+
+	constructor() {
+		this.store = new Map();
+	}
+
+	async read<T>(key: string, setValFn: (key: string) => Promise<CacheElement<T>>): Promise<T> {
+		const cachedValue = this.store.get(key);
+		if (cachedValue !== undefined) {
+			if (cachedValue.expiration > new Date()) {
+				return deserialize(cachedValue.value);
+			}
+			this.store.delete(key);
+		}
+		const val = await setValFn(key);
+		const newCacheValue = { value: serialize(val.value), expiration: val.expiration };
+		this.store.set(key, newCacheValue);
+		return val.value;
+	}
+}
+
+export class APICache implements ReadableCache {
+	constructor(private cacheKey: string) {}
+
+	async read<T>(key: string, setValFn: (key: string) => Promise<CacheElement<T>>): Promise<T> {
+		const cache = await caches.open(this.cacheKey);
+		const request = new Request(`https://${FAKE_DOMAIN_CACHE}/${key}`);
+		const cachedValue = await cache.match(request);
+		if (cachedValue !== undefined) {
+			const val = await cachedValue.text();
+			return deserialize(val);
+		}
+		const val = await setValFn(key);
+		await cache.put(
+			request,
+			new Response(serialize(val.value), {
+				headers: { expires: val.expiration.toUTCString() },
+			})
+		);
+		return val.value;
+	}
+}
+
+export class MixedCache implements ReadableCache {
+	private caches: ReadableCache[];
+	constructor(...caches: ReadableCache[]) {
+		this.caches = caches;
+	}
+
+	async read<T>(key: string, setValFn: (key: string) => Promise<CacheElement<T>>): Promise<T> {
+		const caches = this.caches;
+		const setValFnBuilder = (i: number): ((key: string) => Promise<CacheElement<T>>) => {
+			if (i >= caches.length) {
+				return setValFn;
+			}
+			return async (key: string) => ({
+				value: await caches[i].read(key, setValFnBuilder(i + 1)),
+				expiration: new Date(), // TODO: find a way to use setVal expiration instead
+			});
+		};
+		const cachedValue = await setValFnBuilder(0)(key);
+		return cachedValue.value;
+	}
+}
+
+export const DEFAULT_R2_BUCKET_CACHE_TTL_IN_MS = 5 * 60 * 1000;
+
+export class CachedR2Object {
+	public checksums: R2Checksums;
+	public customMetadata?: Record<string, string>;
+	public etag: string;
+	public httpEtag: string;
+	public httpMetadata?: R2HTTPMetadata;
+	public key: string;
+	public size: number;
+	public uploaded: Date;
+	public version: string;
+
+	constructor(
+		object: R2Object,
+		public body?: Uint8Array
```

Comment:
Can we rename it? At first I thought it was going to be a `body` from a `Response` object which cannot necessarily be read multiple times. I think this is more semantically accurate:
```suggestion
		public data?: Uint8Array
```